### ISSUE TITLE
App integration for shared links

### ISSUE THREAD

**Issue by tomneedham** [owncloud/ios#972]
Discussion topic. I was recently thinking about how nice it can be to be taken to a native app when you access a web link and you have the app installed - although I don't entirely understand the mechanics behind it. However I was wondering if we could do the following:

1. Access shared owncloud link on your mobile device (may have seen it in an email, chat etc)
2. You go to mobile browser
3. Browser / owncloud server pings you to a native app url
4. owncloud app opens up on your mobile device
5. if this points to a file you have in your instance / account on your device, load the files ui and scroll to (internal link)
6. if its not in your owncloud, you can 'add to owncloud' and pick your local account
7. view / open the file using the app if its a public link (open with local applications etc)

This would be a much nicer and native integration than the current experience for shared links on mobile. Maybe you have some other ideas :)

**Comment by michaelstingl** [comment]
From my perspective, this don't work with public links yet. Public links don't get resolved to the location of the files in your account. Private links could work.

Link will open mobile Safaris and ownCloud server should display a "[Smart App Banners](https://developer.apple.com/library/content/documentation/AppleApplications/Reference/SafariWebContent/PromotingAppswithAppBanners/PromotingAppswithAppBanners.html)".

ownCloud iOS app could implement the `application:openURL:sourceApplication:annotation:` method to open the right view inside the app.

Related: https://github.com/owncloud/enterprise/issues/2173 (internal only)

**Comment by michaelstingl** [comment]
Currently, Smart App Banners are broken (tested in 10.0.4):

| ![2017-12-11 12 47 50](https://user-images.githubusercontent.com/214010/33829750-d00a0244-de71-11e7-99f9-f6002f96a55e.png)  | ![2017-12-11 12 48 12](https://user-images.githubusercontent.com/214010/33829751-d024e618-de71-11e7-997e-98ea0d904a8f.png) |
| ------------- | ------------- |
| <img width="798" alt="view-source_https___demo_owncloud_org_login_and_camera_uploads" src="https://user-images.githubusercontent.com/214010/33829915-7a4acfae-de72-11e7-857b-e46c4eeb035c.png">  | <img width="797" alt="view-source_www_math-ninja-app_com_and_camera_uploads" src="https://user-images.githubusercontent.com/214010/33829916-7a629454-de72-11e7-97d2-c44b6dfe0e79.png">  |
| https://demo.owncloud.org/  | http://www.math-ninja-app.com/  |

**Comment by tomneedham** [comment]
>Public links don't get resolved to the location of the files in your account.

Agreed. But I was wondering about an improvement, where you could use the native app to decide what to do (since all the information is public in the link anyway). Like add to owncloud, open the file, share the link elsewhere.


> Private links could work.

And same with this, this would be cool. If we could try the fileid + hostname and see if you can access this file, if not your instance then basically show error, if your instance but no access show forbidden (and future: request access), if your instance but has access scroll to it. That would be awesome integration

**Comment by michaelstingl** [comment]
> But I was wondering about an improvement, where you could use the native app to decide what to do

Please handle this on the server side and unite the different link types. I don't like the idea to solve this on the client side and implement in all 3 client platforms…

**Comment by tomneedham** [comment]
What about 'add to owncloud' button on the web ui opens your native app?

**Comment by michaelstingl** [comment]
No, you need a special URL scheme. We can't register http://mycloud.com/ and http://yourcloud.com/ and http://everycloud.com/ to open the ownCloud iOS app. As far as I know, this string needs to be fixed at time of app compilation.

**Comment by nasli** [comment]
You right @michaelstingl , and this is a save flow

**Comment by nasli** [comment]
The server needs to add a JSON with special configurations and app IDs that allow iOS to open your app. 
We are currently working on it to support open private links and will be available soon

**Comment by tomneedham** [comment]
> No, you need a special URL scheme. We can't register http://mycloud.com/ and http://yourcloud.com/ and http://everycloud.com/ to open the ownCloud iOS app. As far as I know, this string needs to be fixed at time of app compilation.

Ok this makes sense. Is anything actually retrived frmo the url you register, or is it just used to register the app handler? Since the button could point to owncloud.org/ios/publiclink/details.... and ios always registeres to that?

**Comment by nasli** [comment]
>Since the button could point to owncloud.org/ios/publiclink/details.... and ios always registeres to that?

No, @tomneedham If you point to this URL "owncloud.org" needs to have a JSON with the unique app identifier of your apps that you want to open in, add the URL domain inside your app and from that URL been able to retrieve a full path to download the file.

**Comment by nasli** [comment]
This is how universal links works since iOS9 and it's a save approach. iOS no longer support directly URI schemes.

1. You need to set up in your server the appIDs and paths that you support (with an accesible JSON apple-app-site-association)
2. Enable associated domains in your appIDs
3. Add the correct domains in your app capabilities
4. Handle the links in your app

**Comment by SamuAlfageme** [comment]
@michaelstingl about:

> Currently, Smart App Banners are broken (tested in 10.0.3 on iOS 11.2)

I think you're not seeing the banner because at some point in the past you probably dismissed it (and I'm guessing Safari considers all the apps sharing an `app-id` the same app and therefore dismissing all banners that share this value) 

Quoting from Apple's [_Promoting Apps with Smart App Banners_](https://developer.apple.com/library/content/documentation/AppleApplications/Reference/SafariWebContent/PromotingAppswithAppBanners/PromotingAppswithAppBanners.html):

> And with a large and prominent close button, a banner is easy for users to dismiss. **When the user returns to the webpage, the banner won’t reappear.**

Also see https://stackoverflow.com/questions/19412838 on how to bring back Safari to a clean state and bring the banner back.

**Comment by michaelstingl** [comment]
@SamuAlfageme You are right. I clicked away this banner too often…

**Comment by jesmrec** [comment]
This was already shipped in 3.7.4 for public links.

**Comment by michaelstingl** [comment]
> This was already shipped in 3.7.4 for private links.

no solution for unbranded clients yet. I guess this would require a server component to create magic links for iOS users.

**Comment by jesmrec** [comment]
In the server side magic is posible i guess.

iOS users mean App Store users?

But the app would need to be re-signed with specific associated domains in both app and extensions. Distribution via App Store does not allow iOS users to perform it (afaik), so a fix por every user does not seems to be very friendly. Tell me if i am wrong.

CC @pablocarmu

**Comment by michaelstingl** [comment]
proposed server enhancements:
https://github.com/owncloud/core/issues/33337
